rfcs/text/0003_handler_interface.md

```diff
@@ -0,0 +1,299 @@
+   mid-execution. For example, while Elasticsearch clients can technically
+   be replaced throughout the course of the Kibana process, an HTTP route
+   handler should be able to depend on their being a consistent client for its
+   own shorter lifespan.
+
+4. Plugins should not need to pass down high level service contracts throughout
+   their business logic just so they can access them within the context of a
+   handler.
+
+5. Functionality provided by services should not be arbitrarily used in
+   unconstrained execution such as in the plugin lifecycle hooks. For example,
+   it's appropriate for an Elasticsearch client to throw an error if it's used
+   inside an API route and Elasticsearch isn't available, however it's not
+   appropriate for a plugin to throw an error in their start function if
+   Elasticsearch is not available. If the ES client was only made available
+   within the handler context and not to the plugin's start contract at large,
+   then this isn't an issue we'll encounter in the first place.
+
+# Detailed design
+
+There are two parts to this proposal. The first is the handler interface
+itself, and the second is the interface that a service owner implements to make
+their handlers extensible.
+
+## Handler Context
+
+```ts
+interface Context {
+  core: unknown;
+  [contextName: string]: unknown;
+}
+
+type Handler = (context: Partial<Context>, ...args: unknown[]) => Promise<unknown>;
+```
+
+- `args` in this example is specific to the handler type, for instance in a
+  http route handler, this would include the incoming request object.
+- The context object is marked as `Partial<Context>` because the contexts
+  available will vary depending on which plugins are enabled.
+- This type is a convention, not a concrete type. The `core` key should have a
+  known interface that is declared in the service owner's specific Context type.
+
+## Registering new contexts
+
+```ts
+type ContextProvider<T extends keyof Context> = (...args: unknown[]) => Promise<Context[T]>;
+
+interface HandlerService {
+  registerContext<T extends keyof Context>(contextName: T, provider: ContextProvider<T>): void;
+}
+```
+
+- `args` in this example is specific to the handler type, for instance in a http
+  route handler, this would include the incoming request object. It would not
+  include the results from the other context providers in order to keep
+  providers from having dependencies on one another.
+- The `HandlerService` is defined as a literal interface in this document, but
+  in practice this interface is just a guide for the pattern of registering
+  context values. Certain services may have multiple different types of
+  handlers, so they may choose not to use the generic name `registerContext` in
+  favor of something more explicit.
+
+## End to end example
+
+```js
+http.router.registerRequestContext('elasticsearch', async request => {
+  const client = await core.elasticsearch.client$.toPromise();
```

Comment:
it's an implementation detail, but imo important don't forget to make context extension lazy (as a getter?) and execute this function every time when someone reads from `context.[extension point]` otherwise context may have stale version of the client.

Reply:
I think context needs to be generated immediately before invoking the handler. The idea is that context does _not_ change once its corresponding handler begins executing, so even if the ES client were to update at the service level, a handler that has already started executing would continue to use the original client until it finishes executing.

Reply:
@joshdover We should call this out specifically in the RFC because it's an important detail.